### ISSUE TITLE
Validate Truth Catalogs for Run 2: Static and Transient+Variable

### ISSUE THREAD

**Issue by wmwv** [LSSTDESC/DC2-production#370]
Validate the truth catalogs for DC2 Run 2.  
* Static Truth Catalog
* Transient+Variable Truth Catalog

See writeup at this Google Doc for more details on the plan:
https://docs.google.com/document/d/1PnFZqdMmHBk6ePEvZBdKVOt64vwQ1cse2SaSJTVx7GM/view

**Comment by wmwv** [comment]
Static truth catalog distributions compare well with cosmoDC2.
Now investigating percent-level agreement.

**Comment by danielsf** [comment]
@wmwv Do you want me to wait for you to complete this analysis before starting up the remaining 130 healpixels worth of cosmoDC2 truth catalog generation, or did you assume that I had already started that run going?

**Comment by wmwv** [comment]
I think I was ignoring issues about finite time.  I'd like to spend 1 hour more (right now) trying to understand the agreement between the static truth catalog and cosmoDC2 at the 1% level.

**Comment by wmwv** [comment]
@danielsf 
I see ~2% differences in u:

```
print('mag_noMW - ')
print('%4s %12s %12s %12s' % ('band', 'mag', 'mag_true_lsst', 'mag_true_lsst_no_host_extinction'))
bands = ['u', 'g', 'r', 'i', 'z', 'y']
for f in bands:
    print('%-4s' % f,
          '%12.4f' % np.nanmean(combined_df[f'mag_{f}_noMW'] - combined_df[f'mag_{f}_cosmoDC2']),
          '%12.4f' % np.nanmean(combined_df[f'mag_{f}_noMW'] - combined_df[f'mag_true_{f}_lsst']),
          '%12.4f' % np.nanmean(combined_df[f'mag_{f}_noMW'] - combined_df[f'mag_true_{f}_lsst_no_host_extinction']),
         )
```

produces

```
mag_noMW - 
band          mag mag_true_lsst mag_true_lsst_no_host_extinction
u         -0.0262      -0.0352       0.2201
g         -0.0019      -0.0109       0.2060
r          0.0071      -0.0019       0.1988
i          0.0083      -0.0007       0.2038
z          0.0100       0.0010       0.1953
y          0.0019      -0.0072       0.1657
```

If I understand correctly, `mag_noMW - mag_true_lsst` should be the most direct comparison.  I'm fine with the sub-percent level differences, but I'm concerned about the 3% difference in `u`.

See
https://github.com/LSSTDESC/DC2-production/blob/issues/370/Validation/truth_catalog_static.ipynb

**Comment by danielsf** [comment]
Actually, we should be comparing the `_noMW` magnitude to `mag_[ugrizy]_lsst`, since we do want to include the effects of weak gravitational lensing.  That, of course, still gives a 2% effect.

**Comment by danielsf** [comment]
Looking back at the code I used to do the SED fitting, it looks like I actually did fit the magnitudes to `mag_true_*_lsst`, which was probably the wrong thing to do.

(and still doesn't explain the 3% discrepancy in u-band...)

**Comment by wmwv** [comment]
> Looking back at the code I used to do the SED fitting, it looks like I actually did fit the magnitudes to mag_true_*_lsst, which was probably the wrong thing to do.

The lensing should be achromatic, so it shouldn't matter, right?

**Comment by wmwv** [comment]
Ah, or do you mean the SED normalization was fit to the unlensed magnitudes?

**Comment by danielsf** [comment]
I meant that the normalization was fit to the unlensed magnitudes, but, on further inspection, I think that was the *right* thing to do.  ImSim applies lensing magnification as a part of its call to the `galsim.gsobject.GsObject.lens` method here

https://github.com/lsst/sims_GalSimInterface/blob/master/python/lsst/sims/GalSimInterface/galSimInterpreter.py#L411

so we should be passing ImSim something that does not have lensing magnitude already folded into it.

-----

I continued poking at this last night.  I believe the root of the problem is a combination of the following two effects

1) when setting the SED normalization, I used `redshift_true`, which is the cosmological redshift without effects due to galaxy proper motion.  While this is the correct redshift to use when calculating the cosmological distance modulus, we should be using `redshift` (which includes proper motion) when actually redshifting the SED.

2) I was not consistent in whether or not I was using the LSST hardware-only bandpass throughputs, or the hardware-plus-standard-atmosphere throughputs.  Jim and I discussed this on the CosmoSims telecon this morning.  Really, the right thing to do is to use whatever throughput was used in generating the reference catalog that the DM code is calibrating against.  Since ImSim is not yet doing anything to allow the throughput to vary with airmass, it doesn't matter which throughput we use for truth as long as we are consistent between truth, InstanceCatalog, and reference catalog.  I used the hardware-plus-standard-atmosphere throughput when generating the reference catalog, so that is what I will use for the truth catalog going forward.

I need to generate a new SED lookup table and truth catalog that are self consistent.  When I have those ready, I will pass those along for you to test.  Thanks for helping me dig into this.

**Comment by danielsf** [comment]
Comparing the two effects: it appears that using an inconsistent (atmosphere versus no atmosphere) bandpass is the larger of the two problems.

The difference between `redshift` and `redshift_true` results in a few 10**-3 effect on the fluxes.

**Comment by danielsf** [comment]
@wmwv 
I have updated
```
/global/projecta/projectdirs/lsst/groups/SSim/DC2/cosmoDC2_v1.1.4/truth/galaxy_truth_9946.h5
```
I believe this should match cosmoDC2 at the sub-percent level in all bands (comparing `mag_true_*_lsst` to `flux_*_noMW` with the appropriate conversion from ABmags to nanojanskys).  Do you mind rerunning your analysis?

**Comment by wmwv** [comment]
Looks good to me!
```
mag_noMW - 
band          mag mag_true_lsst mag_true_lsst_no_host_extinction
u          0.0091       0.0001       0.2553
g          0.0090      -0.0001       0.2168
r          0.0090      -0.0001       0.2007
i          0.0090      -0.0001       0.2045
z          0.0090      -0.0001       0.1942
y          0.0090      -0.0001       0.1728
```

**Comment by wmwv** [comment]
Branch updated with rendered version.
https://github.com/LSSTDESC/DC2-production/blob/issues/370/Validation/truth_catalog_static.ipynb

**Comment by wmwv** [comment]
@danielsf So galaxies look good.
Should the next step to be to add in the stars?

**Comment by danielsf** [comment]
@wmwv
It is, as always, more complicated than that

The catalog I sent you that exactly matched the cosmoDC2 fluxes was generated by re-fitting the galaxies to SEDs using the hardware+atmosphere bandpasses.  Originally, I had fit the SEDs using the hardware-only bandpasses.  If updating the bandpasses is, in fact, the correct solution, it means that we have to regenerate all of the InstanceCatalogs before starting up Run 2.2i, since the Run 2.1i InstanceCatalogs had SEDs that were fit using hardware-only bandpasses.

This may not be the correct thing to do, though...

cosmoDC2 should, physically, only know about above-the-atmosphere magnitudes (this was the original reasoning behind fitting the SEDs with the hardware-only bandpasses).  It is possible that the actual correct thing to do is to use the old InstanceCatalogs and adjust the DC2 reference catalog so that it contains above-the-atmosphere magnitudes so that we are effectively calibrating out atmospheric extinction.

@jchiang87 and I are discussing this.  There will probably be yet another convocation of experts to decide what the right thing to do is.

**Comment by wmwv** [comment]
What LSST bandpasses did cosmoDC2 use to define its magnitudes?

**Comment by danielsf** [comment]
The hardware-plus-atmosphere bandpasses.

I'm worried, though, that the validation data assumed above-the-atmosphere magnitudes.

When I brought this up at the last CosmoSims telecon, Andrew Hearin was quick to point out that cosmoDC2 was never supposed to be detailed enough to make an above-the-atmosphere vs. below-the-atmosphere distinction.

**Comment by wmwv** [comment]
It's not solely about detail.  There's a significant change in u and g slope due to the atmosphere.

**Comment by wmwv** [comment]
What was used for the stars?

**Comment by danielsf** [comment]
At which point?  When we fit the SEDs to the stars, or when the truth for the stars was generated?

**Comment by wmwv** [comment]
When the truth for the stars was gnerated?

**Comment by evevkovacs** [comment]
@danielsf Why do you think that the validation data assumed above-the-atmosphere magnitudes? When surveys publish data do they normally correct their observed magnitudes for atmospheric effects?
Other points to consider:  most of our validation data uses r and i band (so less affected by this issue). Also much of the validation data in DESCQA is for SDSS  (or has been converted to SDSS). The SDSS magnitudes in cosmoDC2 were obtained by using the SDSS filter definitions available in Galacticus. I would have to check with @abensonca to determine exactly how those were defined.

**Comment by abensonca** [comment]
There are actually two sets of SDSS filters in Galacticus. I'm 99% sure that we used those named, e.g. "SDSS_r" and not those named e.g. "SDSS_r_1.3airmass".

The former is the filter in vacuum (technically, the response curve is the combination of the filter transmission and the CCD response in vacuum). The latter also include an airmass of 1.3.

So, the answer is that these magnitudes from should be above-the-atmosphere.

**Comment by evevkovacs** [comment]
Yes, I confirm that we used SDSS_r and not SDSS_r_1.3airmass etc.

**Comment by danielsf** [comment]
For all of our attempts to convince ourselves that we can fix this with the reference catalog, it looks like we may have to redo the SED fitting, anyway.  This plot shows the change in flux relative to cosmoDC2 that resulted from my confusion about whether to use `redshift` or `redshift_true` when fitting the SED.  The horizontal axes are the cosmoDC2 magnitudes of the galaxies. The vertical axes are log10 of the fractional change in flux.  Vertical lines occur every 2 magnitudes starting at 14 and ending at 26.  As you can see, there are galaxies that will be visible for whom the effect approaches 10**-1.

This means redoing the SED fit and regenerating the InstanceCatalogs, I think.

![dflux_plot](https://user-images.githubusercontent.com/1682854/65644192-8592c900-dfa8-11e9-8cd4-5acc8cb8359d.png)

**Comment by danielsf** [comment]
@evevkovacs: If the validation tests all relied on SDSS data, and Andrew is correct that the bandpasses being used for those magnitudes were above-the-atmosphere bandpasses, them maybe the correct thing to do *is* to fit the SEDs so that, when integrated over the hardware-plus-atmosphere bandpasses, they agree with the `mag_true_*_lsst` columns.

My reasoning:
1) the validation tests imply some agreement between physical truth and the combination of the Galacticus SED and the SDSS hardware bandpasses

2) `mag_true_*_lsst` is a statement about the magnitude in the below-the-atmosphere LSST bandpasses

3) therefore, in order to get an SED that agrees with what we think physical truth is, we should select an SED such that, when integrated with the below-the-atmosphere bandpass, it gives `mag_true_*_lsst`, since (1) implies that the thing being validated is the Galacticus SED.

**Comment by evevkovacs** [comment]
@danielsf The reasoning sounds correct to me. One could also take it one step further and then use the SED to compute _above-atmosphere_ fluxes to feed to the image simulations. This might be confusing however, since then the fluxes in the instance catalogs would not agree exactly with the fluxes in cosmoDC2. We should discuss all this with a wider audience.

**Comment by johannct** [comment]
do we want to redo the fit for the 2.1.1 dataset only, to have a first quick turnaround?

**Comment by danielsf** [comment]
I was going to fit a small area to test, anyway.  I can make it overlap Run 2.1.1.  Good thought.

**Comment by danielsf** [comment]
I am going to summarize here what our plan is so that, when we find ourselves scratching our heads again in six months, we have something to point to.

1) For each galaxy, we will fit a combination of SED, Av, and Rv that most closely matches the colors of the galaxy as reported in cosmoDC2.  Our choice of bandpass (above or below the atmosphere) is somewhat less important in this step because of step (2)

2) For each galaxy, for each bandpass we will find the normalizing magnitude that, in combination with 'redshift' and the (SED, Av, Rv) combination found in step (1), will exactly match the reported value of `mag_true_*_lsst` reported in cosmoDC2 when integrated over the below-the-atmosphere LSST throughputs (airmass=1.2; these are the `baseline/total_*dat` throughputs in the `lsst/throughputs` repo on GitHub).  This means that each galaxy will have one SED, one Av, one Rv, and six magNorms.  There will be a different normalization of the SED, depending on the bandpass in which we are observing the galaxy.  This was the only way that we could match the continuous color distribution of cosmoDC2 galaxies to our grid of SEDs while simultaneously preserving the exact magnitudes reported (and validated) in cosmoDC2.

Note: we are matching `mag_true_*_lsst` in step (2) rather than `mag_*_lsst` because ImSim directly applied magnification due to gravitational lensing, so we need to give ImSim an SED normalized to the unmagnified luminosity of the source, otherwise we will end up double counting the magnification due to lensing.

The code to perform this fitting is being written in the `workspace/sed_cache/` directory of `sims_GCRCatSimInterface`.  Development is being done on the `u/danielsf/sed/190918` branch.

**Comment by wmwv** [comment]
I agree, but Step 1 should be against the below the atmosphere bandpasses.  I agree that it’s a bit less important because of Step 2, but it’s still clearly the correct thing to do.  Our Av and Rv values will be significantly influenced by u and g.  These bands are where both the effects of dust extinction and atmospheric extinction are the greatest.

**Comment by jchiang87** [comment]
I had a very useful discussion with @erykoff yesterday about standard passbands and what our reference catalogs, and by extension, our truth catalogs, should contain in terms of fluxes and magnitudes.

As I understand things, real observations have distinct throughput curves because of differences in airmass from visit to visit, etc..  So in order for the reported fluxes to be most useful, it's desirable to pick a set of standard passbands and apply object SED-dependent chromatic corrections to the observed fluxes to transform them to these standard passbands.  This procedure is described in Burke, et al 2018, AJ, 155, 41.

We can choose our standard passbands at our convenience.  In particular, we can pick passbands that require the least amount of chromatic correction.  Since the DC2 imSim simulations have the same total throughputs in each band for all observations (airmass is always 1.2), we should simply choose those total throughputs as our standard passbands.  As Eli noted in his DESC DE school presentation at Berkeley, this would mean that we wouldn't need to apply any chromatic corrections to the fluxes that come out of the DRP processing.

So, I think the upshot is that we should use the total throughputs (airmass 1.2 + instrumental) everywhere, and specifically in the reference catalogs and truth catalogs.  As long as those throughput curves are available for people to use to interpret the catalog fluxes, that should be just what's needed to make these catalogs useable.

**Comment by wmwv** [comment]
I would be fine with this approach.

The minor bookkeeping bit is that we still have to give the image simulations above-the-atmosphere fluxes.

**Comment by jchiang87** [comment]
The input to imsim is the monochromatic magnitude ("`mag_norm`") at 500nm, i.e., using a delta-function throughput (plus the tabulated SED).  We are determining `mag_norm` by performing band-specific fitting for each visit.  It may be better to use instrument-only throughputs for that, but it's not obvious to me that it is.

**Comment by danielsf** [comment]
This is one of the ways that ImSim differs from PhoSim.

In PhoSim, the input was an SED and a normalization.  PhoSim normalized the SED according to your normalization parameter and then treated that as the above-the-atmosphere SED.  Photons were propagated through the atmosphere and the instrument and the atmosphere-plus-instrument throughput was an emergent property of the simulation.

ImSim takes the same input format, however, ImSim just loads a throughput curve, normalizes the SED, and then integrates the SED over the throughput curve to calculate how many photons to deposit on the silicon.  The throughput loaded by ImSim is the airmass 1.2 throughput from the `lsst/throughputs` repository.  The way ImSim currently works, the only way for atmospheric extinction to get applied is via this loaded throughput curve.  If we really want the fluxes deposited on the silicon to exactly match the contents of `cosmoDC2`, I think this means we have to use below-the-atmosphere fluxes when normalizing the SED (i.e. the normalizing magnitude has to be calculated so that the below-the-atmosphere magnitude matches the `mag_true_*_lsst` columns in `cosmoDC2`).

Maybe there was no contention on this point, but I thought Michael was just advocating for something different than what I just described and felt it was important to explain this difference between PhoSim and ImSim.

**Comment by evevkovacs** [comment]
Note also that the throughputs should match the version used for cosmoDC2. These are tagged in the LSST throughputs repository and are different from the latest version.

**Comment by wmwv** [comment]
Thanks @danielsf for the explanation above.  I am both now better informed and happy with the proposal.

**Comment by danielsf** [comment]
I do not believe I was as precise as I needed to be when stating our plan above.  This is a more accurate description of how the SED fitting takes place (I am leaving the original comment with a caveat at the end so that, when we are looking back at this conversation, we can understand the context of the conversation that followed)

The problem that we are trying to solve is as follows:

For historical reasons, ImSim expects each source to be input as a detailed SED, a normalizing magnitude, and two parameters (Av, Rv) specifying the extinction due to internal dust (we are using the Cardelli, Clayton, and Mathis dust extinction law).  We are working with a library of 959 Bruzual and Charlot galaxy SEDs assembled by the LSST Project simulations team.  Obviously, this grid of SEDs is incapable of exactly reproducing the continuous distribution of galaxy colors seen in cosmoDC2.  Therefore, we have decided that each galaxy will be characterized by the parameters (SED_disk, Av_disk, Rv_disk, magNorm_disk, SED_bulge, Av_bulge, Rv_bulge, magNorm_bulge) where SED_*, Av_*, Rv_* have one value per galaxy per component (component = "bulge" or "disk") and magNorm_* has six values per galaxy per component, one for each LSST bandpass.  We will adjust the magNorm on a per-band basis so as to exactly reproduce the LSST colors in cosmoDC2.

The code to perform this fitting is being written in the `workspace/sed_cache/` directory of `sims_GCRCatSimInterface`.  Development is being done on the `u/danielsf/sed/190918` branch.

The specific steps to do this fitting are

1) For each galaxy component (disk or bulge), find the combination of SED, Av, Rv that best reproduces the colors implied reported by the fluxes listed in the 30 tophat filters that cosmoDC2 uses to approximate the SED of each galaxy component (these columns have names like `sed_1000_246_disk` in cosmoDC2).

1a) For each bandpass, find the magNorm that, given the SED, Av, Rv combination fit above, best reproduces the observed per-component fluxes reported by cosmoDC2 (i.e. in `LSST_filters/diskLuminositiesStellar:LSST_g:observed:dustAtlas`).  Note: for this calculation, be sure to use the `redshift` parameter, rather than `redshift_true`, since the former includes the effect of galaxy proper motions.

Steps (1) and (1a) above are accomplished by the method `sed_from_galacticus_mags` in `workspace/sed_cache/SedFitter.py` in `sims_GCRCatSimInterface`.

At this point, we have all of the parameters we said we would solve for, however, we find that the resulting observed LSST magnitudes of the entire galaxies do not yet exactly match the `mag_true_*_lsst` columns reported in cosmoDC2.  Therefore, we make one more pass at fitting

2) For each galaxy as a whole (bulge and disk)
2a) For each bandpass, find the `delta_magnitude` which, wen added to `magNorm_disk` and `magNorm_bulge` for that bandpass brings the entire galaxy's magnitude into near perfect agreement with `mag_true_*_lsst`.  Add this `delta_magnitude` to the appropriate `magNorm_disk` and `magNorm_flux` values.  This step gives the final values of the 12 magNorms per galaxy (6 for the galaxy's disk; 6 for the galaxy's bulge)

Note: we are matching `mag_true_*_lsst` in step (2) rather than `mag_*_lsst` because ImSim directly applied magnification due to gravitational lensing, so we need to give ImSim an SED normalized to the unmagnified luminosity of the source, otherwise we will end up double counting the magnification due to lensing.

**Comment by danielsf** [comment]
It appears that we may have done the right thing in spite ourselves (except that I still screwed up by fitting the SEDs through the hardware-only bandpass).  The following is cut-and-pasted from a Slack conversation I had with Lynne Jones (this is all from Lynne, who actually knows what she is talking about):

---
So: the catalogs DO report “above-atmosphere” magnitudes. However, there is still a ‘shape’ to that throughput curve, and it is indeed the shape of a throughput curve that includes atmosphere - an X=1.2 atmosphere, to be exact.

So you did the right thing - you should calculate magnitudes using the standard ‘total throughput’ curves, including the X=1.2 atmosphere. But you should set the *zeropoint* appropriately for an AB=0 star to have 3631 Jy flux in that same passband.

If you calculate magnitudes using Sed, this is done properly. [Scott: Sed in this context refers to a class in the lsst_sims stack, which is what we do use for calculating the magnitude of sources given an SED and a bandpass]

If you calculate flux, and then convert to magnitude using some other zeropoint, then you may or may not end up at the same point.

But, yes, you should include the atmosphere in that calculation. (For a useful thought experiment, think of what would happen in u band if you did not include the atmosphere … a very blue star would have quite a different magnitude with or without the atmosphere, as most of the cutoff in throughput comes from the atmosphere).

...

Anyway - what I can say, is that EVERY project reports “above-atmosphere” magnitudes. Because everybody zeropoints to what a star would have if there was no atmospheric extinction.

And yet every project also generates these zeropoints using a complete throughput curve that includes some standard or average atmosphere and hardware throughput function.

So there is a theoretical white dwarf SED above the atmosphere, but that is not what anyone reports. They always report “this is the flux that we measured, in our bandpass that includes an average atmosphere and hardware, and we think that this star / other stars we’re using as zeropoints / should generate X photons per second above the atmosphere, and this one generated Y photons, so here’s its calibrated magnitude” (at which point, it’s a calibrated, above-atmosphere, in-the-standard-passband magnitude).

Now, sky is special — because it *is* actually reported as “at the telescope pupil” flux, so if you’re calculating magnitudes for the sky, you should just use hardware. But for everything else, it should be the hardware + atmosphere.

-----

most of what Lynne is referring to should, in principal, be documented in http://ls.st/lse-180*
